{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# CSF ANALYSIS FOR PPMI data\n",
    "\n",
    "\n",
    "Created: 2018-08-07    \n",
    "By: Hirotaka Iwaki\n",
    "\n",
    "Required data for this analysis:    \n",
    "* \"Current_Biospecimen_Analysis_Results.csv\" from PPMI-LONI    \n",
    "* \"01_extract_EVENTDATE.csv\" derived from PPMI dataset    \n",
    "* \"02_LatestDiagDemog.csv\" derived from PPMI dataset: ([codes](https://github.com/hirotaka-i/PDcohorts/blob/master/PPMI/codes))   \n",
    "* Genotypings data from LNG: Imputed and filtered (MAF >0.001, Rsq >0.3, and cut per 20K SNPs per file.)\n",
    "* \"PPMI.pca.eigenvec\": PCs derived from genotyping data\n",
    "\n",
    "\n",
    "## Objective/Model\n",
    "GWAS analysis for CSF $\\alpha$-synuclein concentrations.    \n",
    "A linear model adjusting for sex, age, and CSF hemoglobin >200 with random intercept for individuals.     \n",
    "After diagnosis, the CSF $\\alpha$-synuclein concentrations don't change much. -> no need to think about the effect on slope.    \n",
    "Patients and healthy controls are analyzed separately and meta-analized if appropriate.\n",
    "\n",
    "## Data preparation\n",
    "### CSF file\n",
    "Extract $\\alpha$-synuclein concentrations and CSF hemoglobin levels from \"Current_Biospecimen_Analysis_Results.csv\" and give the visiting date from \"01_extract_EVENTDATE.csv\"."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "mkdir -p temp\n",
    "awk 'BEGIN{FS=\",\";OFS=\"\\t\"}/CSF Alpha-synuclein/{print $1\"_\"$4,$7}' \\\n",
    "    data/Current_Biospecimen_Analysis_Results.csv | \\\n",
    "    sed 's/\"//g' | sort > temp/asyn.txt\n",
    "awk 'BEGIN{FS=\",\";OFS=\"\\t\"}/CSF Hemoglobin/{print $1\"_\"$4,$7}' \\\n",
    "    data/Current_Biospecimen_Analysis_Results.csv | \\\n",
    "    sed 's/\"//g' | sort > temp/hemo.txt\n",
    "awk 'BEGIN{FS=\",\";OFS=\"\\t\"}//{print $1\"_\"$2,$3}' data/01_extract_EVENTDATE.csv | \\\n",
    "    sed 's/\"//g' | sort > temp/event_date.txt\n",
    "join temp/asyn.txt temp/hemo.txt | join - temp/event_date.txt | sed 's/_/ /g' > \\\n",
    "    temp/asyn_hemo.txt # space separated"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Merge the above  file with Demographic file\n",
    "merge with \"02_LatestDiagDemog.csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "echo '\n",
    "library(dplyr)\n",
    "library(data.table)\n",
    "t1 = fread(\"data/02_LatestDiagDemog.csv\", header=T)\n",
    "t2 = fread(\"temp/asyn_hemo.txt\", header=F, , na.strings=\"N/A\") %>% \n",
    "    mutate(V1 = as.character(V1)) %>% #  V1(ID) regarded as int\n",
    "    filter(!is.na(V2) & !is.na(V3)) # delete \"N/A\" value observation in ANSN and HEMO\n",
    "names(t2)=c(\"ID\", \"EVENT_ID\",\"ASYN\", \"HEMO\", \"DATE\")\n",
    "d1 = left_join(t2, t1, by = \"ID\")\n",
    "d1 %>% mutate(HEMO_num = as.numeric(HEMO)) %>% filter(is.na(HEMO_num)) %>%\n",
    "    with(table(HEMO, useNA=\"always\"))\n",
    "##############################\n",
    "# The output of above line ###########\n",
    "# HEMO\n",
    "#   >20 above below  <NA>\n",
    "#    28   314  1918     0\n",
    "# Warning message:\n",
    "# In evalq(as.numeric(HEMO), <environment>) : NAs introduced by coercion ((<- Warning but OK))\n",
    "# ((Although, I suspect \">20\" is actually \"<20\", I will eliminate \">20\" obs in the following))\n",
    "###############################\n",
    "d2 = d1 %>% mutate(HEMO200 = case_when(\n",
    "    HEMO==\">20\"~99, # for elimination\n",
    "    HEMO==\"below\"~ 0,\n",
    "    HEMO==\"above\"~ 1,\n",
    "    TRUE~ if_else(as.numeric(HEMO)>200, 1, 0))) %>%\n",
    "    mutate(logASYN = log(ASYN),\n",
    "           AGE = (as.numeric(as.Date(DATE)) - BIRTHDT)/365.25) %>% \n",
    "    filter(HEMO200!=99) %>% # eliminate HEMO = \">20\" obs\n",
    "    mutate(CASE = case_when(\n",
    "      grepl(\"PD_YsImg_PPMI\", DIAG) ~ 1, \n",
    "      grepl(\"CTR_YsImg_PPMI\", DIAG) ~ 0,\n",
    "      TRUE ~ 99\n",
    "    )) %>%\n",
    "    mutate (IID = paste(ID,ID,sep=\"_\")) # for later step to merge with Genotyping data\n",
    "d2 %>% with(table(CASE), useNA=\"always\")\n",
    "d3 = d2 %>% filter(CASE!=99) %>% # num_of_obs 3200 -> 2600\n",
    "    select(IID,ASYN,CASE, FEMALE,AGE,HEMO200) %>%\n",
    "    arrange(IID)\n",
    "write.table(d3, \"temp/pheno.txt\", row.names = F, quote = F, sep = \"\\t\")\n",
    "print(\"complete\")\n",
    "' > _pheno.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HEMO\n",
      "  >20 above below  <NA> \n",
      "   28   314  1917     0 \n",
      "CASE\n",
      "   0    1   99 \n",
      " 797 1747  663 \n",
      "[1] \"complete\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[+] Loading gcc  7.2.0  ... \n",
      "[+] Loading GSL 2.4 for GCC 7.2.0 ... \n",
      "[+] Loading openmpi 3.0.0  for GCC 7.2.0 \n",
      "[+] Loading R 3.5.0_build2 \n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "Attaching package: ‘data.table’\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "Warning message:\n",
      "In evalq(as.numeric(HEMO), <environment>) : NAs introduced by coercion\n",
      "Warning message:\n",
      "In if_else(as.numeric(HEMO) > 200, 1, 0) : NAs introduced by coercion\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "module load R\n",
    "Rscript --vanilla _pheno.R"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Mege with the PCs file\n",
    "merge with \"PPMI.pca.eigenvec\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "sort temp/pheno.txt > temp/pheno_sort.txt\n",
    "cut -f 2- data/PPMI.pca.eigenvec | sort | join -t$'\\t' temp/pheno_sort.txt -  > temp/pheno_PC5_temp.txt\n",
    "grep 'IID' temp/pheno_PC5_temp.txt > temp/pheno_PC5_colnames.txt\n",
    "grep -v 'IID' temp/pheno_PC5_temp.txt | cat temp/pheno_PC5_colnames.txt - > temp/pheno_PC5.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## GLMM analysis\n",
    "\n",
    "### R code for analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "echo '\n",
    "args <- commandArgs(trailingOnly = TRUE)\n",
    "OUTCOME = args[1]\n",
    "DATASET = args[2]\n",
    "chrNUM =  args[3]\n",
    "ITER =  args[4]\n",
    "FOLDER = args[5]\n",
    "CASECODE = as.numeric(args[6])\n",
    "\n",
    "library(data.table);library(dplyr);library(lme4)\n",
    "# Read pheno.txt and set COV (PC1 and PC2)\n",
    "cohort=fread(\"temp/pheno_PC5.txt\") # %>% filter(CASE==CASECODE) # Only limiting to cases\n",
    "COVs = names(cohort)[c(-(1:3),-9, -10, -11)] # eliminate ID, OUTCOME, CASE, PC3-5 from covariates\n",
    "# Read genotyping data\n",
    "SNPset = fread(paste(\"zcat -f \", FOLDER, \"/\", DATASET, \"/\", chrNUM, \".\", ITER, \".trans.txt.gz\", sep=\"\"))\n",
    "SNPs = names(SNPset)[-(1:2)] # 1 ID, 2 DOSE, SNP name starts from 3\n",
    "## Merge\n",
    "cohort_snp = left_join(cohort, SNPset, by = c(\"IID\"=\"ID\"))\n",
    "write.csv(cohort_snp[,1:20], \"temp/_temp.csv\")\n",
    "# Set function for analysis\n",
    "glmm.listfunc = function(x){\n",
    "  # Models\n",
    "  MODEL = paste(OUTCOME, \"~\", \"`\", SNPs[x], \"`+\", paste(COVs, collapse=\"+\"), \"+(1|IID)\", sep = \"\")\n",
    "  testLmer = try(lmer(eval(parse(text = MODEL)), data = cohort_snp),silent = T)\n",
    "  if(class(testLmer)[1]==\"try-error\"){\n",
    "    sumstat=rep(\"DROP\",6)\n",
    "  }else{\n",
    "    temp = summary(testLmer)\n",
    "    temp1 = temp$coefficients\n",
    "    if(grep(SNPs[x], rownames(temp1)) %>% length == 0){ # In this case, SNP is dropeed from the model\n",
    "      sumstat=rep(NA,6)\n",
    "    }else{\n",
    "      RES = temp1[2,] # The first row is intercept\n",
    "      PV_APPROX = 2 * pnorm(abs(RES[3]), lower.tail=F) # df is enough large for approximation\n",
    "      OBS_N = paste(length(temp$residuals), \"_\", temp$ngrps, sep=\"\")\n",
    "      sumstat <- c(SNPs[x], OBS_N, RES[3], RES[1], RES[2], PV_APPROX)\n",
    "    }\n",
    "  }\n",
    "  return(sumstat)\n",
    "}\n",
    "temp = lapply(1:length(SNPs), glmm.listfunc)\n",
    "temp2 = do.call(rbind, temp)\n",
    "attributes(temp2)$dimnames[[2]]=c(\"SNP\", \"OBS_N\", \"Tvalue\", \"Beta\", \"SE\", \"Pv_approx\")\n",
    "head(temp2)\n",
    "NEWDIR = paste(\"/data/LNG/Hirotaka/ASYN/\", OUTCOME, \"/\", chrNUM, sep = \"\")\n",
    "dir.create(NEWDIR, recursive = T, showWarnings = F)\n",
    "write.table(temp2, paste(NEWDIR, \"/\", DATASET, CASECODE, \".\", ITER, \".txt\", sep=\"\"), row.names = F, quote = F, sep = \"\\t\")\n",
    "print(\"complete\")\n",
    "' > _analysis.R"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "testrun"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     SNP                OBS_N      Tvalue               Beta               \n",
      "[1,] \"12:133435735_C_T\" \"2211_499\" \"0.466716837974515\"  \"299.734360406519\" \n",
      "[2,] \"12:133435793_C_T\" \"2211_499\" \"-0.647698843465301\" \"-555.989155686094\"\n",
      "[3,] \"12:133435861_C_T\" \"2211_499\" \"1.14089221530962\"   \"108.715849628715\" \n",
      "[4,] \"12:133436127_A_G\" \"2211_499\" \"1.02662554369804\"   \"239.823438043981\" \n",
      "[5,] \"12:133436499_C_T\" \"2211_499\" \"0.0820380234755894\" \"5.38176923817528\" \n",
      "[6,] \"12:133437240_G_A\" \"2211_499\" \"-0.668619623065455\" \"-30.8986974886041\"\n",
      "     SE                 Pv_approx          \n",
      "[1,] \"642.218870241158\" \"0.640702481297696\"\n",
      "[2,] \"858.406898970911\" \"0.517179752422419\"\n",
      "[3,] \"95.2902019751371\" \"0.253914779350272\"\n",
      "[4,] \"233.60361479036\"  \"0.304596821289577\"\n",
      "[5,] \"65.6009129690532\" \"0.934616476986448\"\n",
      "[6,] \"46.212669240758\"  \"0.503738151738932\"\n",
      "[1] \"complete\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[+] Loading gcc  7.2.0  ... \n",
      "[+] Loading GSL 2.4 for GCC 7.2.0 ... \n",
      "[+] Loading openmpi 3.0.0  for GCC 7.2.0 \n",
      "[+] Loading R 3.5.0_build2 \n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "The following objects are masked from ‘package:data.table’:\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "Loading required package: Matrix\n",
      "Warning messages:\n",
      "1: Some predictor variables are on very different scales: consider rescaling \n",
      "2: Some predictor variables are on very different scales: consider rescaling \n",
      "3: Some predictor variables are on very different scales: consider rescaling \n",
      "4: Some predictor variables are on very different scales: consider rescaling \n",
      "5: Some predictor variables are on very different scales: consider rescaling \n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "module load R\n",
    "Rscript --vanilla _analysis.R ASYN PPMI chr12 11  /data/LNG/Hirotaka/progGWAS/SNPfilter/maf001rsq3_20Kcut 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create .swarm file and submit\n",
    "\"swirm\" is an NIH-wrapper for job array    \n",
    "create .swirm files and then submit these jobs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "FOLDER=/data/LNG/Hirotaka/progGWAS/SNPfilter/maf001rsq3_20Kcut\n",
    "rm _analysis*.swarm \n",
    "for j in $(ls $FOLDER/PPMI);do\n",
    "    chrNUM=$(echo $j | cut -d '.' -f 1)\n",
    "    ITER=$(echo $j | cut -d '.' -f 2)\n",
    "    for CASECODE in 0 1;do\n",
    "        if [ $CASECODE == 1 ]; then\n",
    "            echo \"Rscript --vanilla _analysis.R ASYN PPMI $chrNUM $ITER  $FOLDER $CASECODE\" >> _analysisCASE.swarm\n",
    "        else\n",
    "            echo \"Rscript --vanilla _analysis.R ASYN PPMI $chrNUM $ITER  $FOLDER $CASECODE\" >> _analysisCTRL.swarm\n",
    "        fi\n",
    "    done;\n",
    "done;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading modules R\n",
      "207 commands run in 13 subjobs, each command requiring 3 gb and 1 thread, packing 2 processes per subjob, running 8 processes serially per subjob\n",
      "sbatch --array=0-12 --job-name=swarm --output=/dev/null --error=/dev/null --cpus-per-task=2 --mem=6144 --partition=norm --time=24:00:00 /spin1/swarm/iwakih2/S21ST0WsVw/swarm.batch\n",
      "Loading modules R\n",
      "207 commands run in 13 subjobs, each command requiring 3 gb and 1 thread, packing 2 processes per subjob, running 8 processes serially per subjob\n",
      "sbatch --array=0-12 --job-name=swarm --output=/dev/null --error=/dev/null --cpus-per-task=2 --mem=6144 --partition=norm --time=24:00:00 /spin1/swarm/iwakih2/iDRkch3h5K/swarm.batch\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "rm -rf swarm_analysis\n",
    "swarm -f  _analysisCASE.swarm --time=3:00:00 -g 3 -p 2 -b 8 --logdir ./swarm_CASE --module R --devel\n",
    "swarm -f  _analysisCTRL.swarm --time=3:00:00 -g 3 -p 2 -b 8 --logdir ./swarm_CTRL --module R --devel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "6811357"
      ]
     },
     "execution_count": 155,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "6811355\n",
    "6811357"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## meta_analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 179,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.axes._subplots.AxesSubplot at 0x2aaaf84b1b00>"
      ]
     },
     "execution_count": 179,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYAAAAD8CAYAAAB+UHOxAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvhp/UCwAAGVlJREFUeJzt3X+QHOV95/H3J5L5ubYkkNlTJFVWXBQuBPlsaU/I8ZVrFsVIYBeiUqZOlM5IjlyqSzBHAi4jzpWj8oMyzpmQUPGR6ILOkDisCSZBJclHFMHGxVUQIIyRhIy1BhWsJCMTCeUW/CNyvvdHP2tGy6x2dmY13arn86qamu6nn+nnMzMtfbe7Z6YVEZiZWX5+puwAZmZWDhcAM7NMuQCYmWXKBcDMLFMuAGZmmXIBMDPLlAuAmVmmXADMzDLlAmBmlqmpZQc4mZkzZ0ZPT0/DZW+++SbnnntuZwONw5maV8VcztScKmaCauYqK9POnTtfj4j3jtsxIip7W7RoUYzl8ccfH3NZWZypeVXM5UzNqWKmiGrmKisT8Ew08X+sDwGZmWXKBcDMLFMuAGZmmXIBMDPLlAuAmVmmXADMzDLlAmBmlikXADOzTLkAmJllqtI/BXG66lm/pbSx99/x0dLGNrPTi/cAzMwy5QJgZpYpFwAzs0y5AJiZZWrcAiBpo6TDknaPar9B0ouS9kj6g7r2WyUNpmXL6tqXp7ZBSesn92mYmdlENfMpoC8DfwLcP9IgqQ9YAbwvIn4k6YLUfjGwEvgl4GeBv5f0C+lhXwI+AgwBT0vaFBEvTNYTMTOziRm3AETENyT1jGr+deCOiPhR6nM4ta8A+lP7y5IGgcVp2WBEvAQgqT/1dQEwMyuJiovHjNOpKACbI+KSNP8c8AiwHPgh8JmIeFrSnwBPRsRfpn73Al9Pq1keEZ9K7Z8ALo2ITzcYax2wDqC7u3tRf39/w0zDw8N0dXU1/0w7YCTTrgPHSsuwYPa0E+ar+DpBNXM5U3OqmAmqmausTH19fTsjone8fq1+EWwqMANYAvwH4EFJFwJq0DdofK6hYeWJiA3ABoDe3t6o1WoNAwwMDDDWsrKMZFpT5hfBVtVOmK/i6wTVzOVMzaliJqhmripmqtdqARgCHk7XnnxK0r8CM1P73Lp+c4CDaXqsdjMzK0GrHwP9W+AygHSS9wzgdWATsFLSmZLmAfOBp4CngfmS5kk6g+JE8aZ2w5uZWevG3QOQ9ABQA2ZKGgJuAzYCG9NHQ38MrE57A3skPUhxcvc4cH1E/CSt59PAo8AUYGNE7DkFz8fMzJrUzKeArh1j0X8eo//twO0N2rcCWyeUzszMThl/E9jMLFMuAGZmmXIBMDPLlAuAmVmmXADMzDLlAmBmlikXADOzTLkAmJllygXAzCxTLgBmZplyATAzy5QLgJlZplwAzMwy5QJgZpYpFwAzs0y5AJiZZWrcAiBpo6TD6epfo5d9RlJImpnmJeluSYOSnpe0sK7vakn70m315D4NMzObqGb2AL4MLB/dKGku8BHglbrmKyiuAzwfWAfck/qeR3EpyUuBxcBtkma0E9zMzNozbgGIiG8ARxosugv4LBB1bSuA+6PwJDBd0ixgGbAtIo5ExFFgGw2KipmZdY6Ka7mP00nqATZHxCVp/ipgaUTcKGk/0BsRr0vaDNwREU+kftuBWyguKn9WRPx+av9t4AcR8cUGY62j2Hugu7t7UX9/f8NMw8PDdHV1TejJnmojmXYdOFZahgWzp50wX8XXCaqZy5maU8VMUM1cZWXq6+vbGRG94/Ub96Lwo0k6B/gccHmjxQ3a4iTt72yM2ABsAOjt7Y1ardYwx8DAAGMtK8tIpjXrt5SWYf+q2gnzVXydoJq5nKk5VcwE1cxVxUz1WvkU0L8F5gHfSn/9zwGelfRvgCFgbl3fOcDBk7SbmVlJJlwAImJXRFwQET0R0UPxn/vCiPgesAm4Ln0aaAlwLCIOAY8Cl0uakU7+Xp7azMysJM18DPQB4B+BiyQNSVp7ku5bgZeAQeB/Ab8BEBFHgN8Dnk63301tZmZWknHPAUTEteMs76mbDuD6MfptBDZOMJ+ZmZ0i/iawmVmmXADMzDLlAmBmlikXADOzTLkAmJllygXAzCxTLgBmZplyATAzy5QLgJlZplwAzMwy5QJgZpYpFwAzs0y5AJiZZcoFwMwsUy4AZmaZauaCMBslHZa0u67tf0j6tqTnJf2NpOl1y26VNCjpRUnL6tqXp7ZBSesn/6mYmdlENLMH8GVg+ai2bcAlEfE+4DvArQCSLgZWAr+UHvM/JU2RNAX4EnAFcDFwbeprZmYlGbcARMQ3gCOj2v4uIo6n2ScpLvIOsALoj4gfRcTLFJeGXJxugxHxUkT8GOhPfc3MrCSTcQ7g14Cvp+nZwKt1y4ZS21jtZmZWEhWX8R2nk9QDbI6IS0a1fw7oBX41IkLSl4B/jIi/TMvvpbhQ/M8AyyLiU6n9E8DiiLihwVjrgHUA3d3di/r7+xtmGh4epqurq8mn2RkjmXYdOFZahgWzp50wX8XXCaqZy5maU8VMUM1cZWXq6+vbGRG94/Ub96LwY5G0GvgYsDTeriJDwNy6bnOAg2l6rPYTRMQGYANAb29v1Gq1huMPDAww1rKyjGRas35LaRn2r6qdMF/F1wmqmcuZmlPFTFDNXFXMVK+lQ0CSlgO3AFdFxFt1izYBKyWdKWkeMB94CngamC9pnqQzKE4Ub2ovupmZtWPcPQBJDwA1YKakIeA2ik/9nAlskwTwZET8l4jYI+lB4AXgOHB9RPwkrefTwKPAFGBjROw5Bc/HzMyaNG4BiIhrGzTfe5L+twO3N2jfSnE+wMzMKsDfBDYzy5QLgJlZplwAzMwy5QJgZpYpFwAzs0y5AJiZZcoFwMwsUy4AZmaZcgEwM8uUC4CZWaZcAMzMMuUCYGaWKRcAM7NMuQCYmWXKBcDMLFMuAGZmmRq3AEjaKOmwpN11bedJ2iZpX7qfkdol6W5Jg5Kel7Sw7jGrU/996XrCZmZWomb2AL4MLB/Vth7YHhHzge1pHuAKiusAzwfWAfdAUTAoLiV5KbAYuG2kaJiZWTnGLQAR8Q3gyKjmFcB9afo+4Oq69vuj8CQwXdIsYBmwLSKORMRRYBvvLCpmZtZBrZ4D6I6IQwDp/oLUPht4ta7fUGobq93MzEqiiBi/k9QDbI6IS9L8GxExvW750YiYIWkL8PmIeCK1bwc+C1wGnBkRv5/afxt4KyLubDDWOorDR3R3dy/q7+9vmGl4eJiurq4JPNVTbyTTrgPHSsuwYPa0E+ar+DpBNXM5U3OqmAmqmausTH19fTsjone8flNbXP9rkmZFxKF0iOdwah8C5tb1mwMcTO21Ue0DjVYcERuADQC9vb1Rq9UadWNgYICxlpVlJNOa9VtKy7B/Ve2E+Sq+TlDNXM7UnCpmgmrmqmKmeq0eAtoEjHySZzXwSF37denTQEuAY+kQ0aPA5ZJmpJO/l6c2MzMrybh7AJIeoPjrfaakIYpP89wBPChpLfAKcE3qvhW4EhgE3gI+CRARRyT9HvB06ve7ETH6xLKZmXXQuAUgIq4dY9HSBn0DuH6M9WwENk4onZmZnTL+JrCZWaZcAMzMMuUCYGaWKRcAM7NMuQCYmWXKBcDMLFMuAGZmmXIBMDPLlAuAmVmmXADMzDLlAmBmlikXADOzTLkAmJllygXAzCxTLgBmZplyATAzy1RbBUDSb0naI2m3pAcknSVpnqQdkvZJ+qqkM1LfM9P8YFreMxlPwMzMWtNyAZA0G/ivQG9EXAJMAVYCXwDuioj5wFFgbXrIWuBoRPw8cFfqZ2ZmJWn3ENBU4GxJU4FzgEPAZcBDafl9wNVpekWaJy1fKkltjm9mZi1quQBExAHgixQXhT8EHAN2Am9ExPHUbQiYnaZnA6+mxx5P/c9vdXwzM2uPiuu4t/BAaQbwNeA/AW8Af53mb0uHeZA0F9gaEQsk7QGWRcRQWvZdYHFE/NOo9a4D1gF0d3cv6u/vbzj+8PAwXV1dLWU/VUYy7TpwrLQMC2ZPO2G+iq8TVDOXMzWnipmgmrnKytTX17czInrH6ze1jTF+BXg5Ir4PIOlh4JeB6ZKmpr/y5wAHU/8hYC4wlA4ZTQOOjF5pRGwANgD09vZGrVZrOPjAwABjLRvRs37LhJ9UO25e8BPufOJN2ntZ27N/Ve2E+WZepzJUMZczNaeKmaCauaqYqV475wBeAZZIOicdy18KvAA8Dnw89VkNPJKmN6V50vLHotXdDzMza1s75wB2UJzMfRbYlda1AbgFuEnSIMUx/nvTQ+4Fzk/tNwHr28htZmZtautYRUTcBtw2qvklYHGDvj8ErmlnPDMzmzz+JrCZWaZcAMzMMuUCYGaWKRcAM7NMuQCYmWXKBcDMLFMuAGZmmXIBMDPLlAuAmVmmXADMzDLlAmBmlikXADOzTLkAmJllygXAzCxTLgBmZplyATAzy1RbBUDSdEkPSfq2pL2SPijpPEnbJO1L9zNSX0m6W9KgpOclLZycp2BmZq1odw/gj4H/ExH/Dvj3wF6KSz1uj4j5wHbevvTjFcD8dFsH3NPm2GZm1oaWC4Ck9wAfJl3zNyJ+HBFvACuA+1K3+4Cr0/QK4P4oPAlMlzSr5eRmZtYWRURrD5TeT3ER+Bco/vrfCdwIHIiI6XX9jkbEDEmbgTsi4onUvh24JSKeGbXedRR7CHR3dy/q7+9vOP7w8DBdXV0nzbjrwLGWnlurus+G137Q0SHfYcHsaSfMN/M6laGKuZypOVXMBNXMVVamvr6+nRHRO16/di4KPxVYCNwQETsk/TFvH+5pRA3a3lF9ImIDRWGht7c3arVaw5UNDAww1rIRa9ZvOenyyXbzguPcuaudl7R9+1fVTphv5nUqQxVzOVNzqpgJqpmripnqtXMOYAgYiogdaf4hioLw2sihnXR/uK7/3LrHzwEOtjG+mZm1oeUCEBHfA16VdFFqWkpxOGgTsDq1rQYeSdObgOvSp4GWAMci4lCr45uZWXvaPV5xA/AVSWcALwGfpCgqD0paC7wCXJP6bgWuBAaBt1Jfm2Q9ow573bzgeEcOhe2/46OnfAwzm1xtFYCIeA5odKJhaYO+AVzfznhmZjZ5/E1gM7NMuQCYmWXKBcDMLFMuAGZmmXIBMDPLlAuAmVmmXADMzDLlAmBmlikXADOzTLkAmJllygXAzCxTLgBmZplyATAzy5QLgJlZplwAzMwy5QJgZpaptguApCmSvilpc5qfJ2mHpH2SvpquFoakM9P8YFre0+7YZmbWusnYA7gR2Fs3/wXgroiYDxwF1qb2tcDRiPh54K7Uz8zMStJWAZA0B/go8OdpXsBlwEOpy33A1Wl6RZonLV+a+puZWQlUXKq3xQdLDwGfB94NfAZYAzyZ/spH0lzg6xFxiaTdwPKIGErLvgtcGhGvj1rnOmAdQHd396L+/v6GYw8PD9PV1XXSfLsOHGv5ubWi+2x47QcdHXJcncq0YPa0CfVv5v3rNGdqThUzQTVzlZWpr69vZ0Q0ul77CVq+KLykjwGHI2KnpNpIc4Ou0cSytxsiNgAbAHp7e6NWq43uAsDAwABjLRuxZv2Wky6fbDcvOM6du1p+SU+JTmXav6o2of7NvH+d5kzNqWImqGauKmaq187/DB8CrpJ0JXAW8B7gj4DpkqZGxHFgDnAw9R8C5gJDkqYC04AjbYxvZmZtaPkcQETcGhFzIqIHWAk8FhGrgMeBj6duq4FH0vSmNE9a/li0c/zJzMzaciq+B3ALcJOkQeB84N7Ufi9wfmq/CVh/CsY2M7MmTcrB4YgYAAbS9EvA4gZ9fghcMxnjmZlZ+/xNYDOzTLkAmJllygXAzCxTLgBmZplyATAzy5QLgJlZplwAzMwy5QJgZpYpFwAzs0y5AJiZZcoFwMwsUy4AZmaZcgEwM8uUC4CZWaZcAMzMMtVyAZA0V9LjkvZK2iPpxtR+nqRtkval+xmpXZLuljQo6XlJCyfrSZiZ2cS1swdwHLg5In4RWAJcL+liiit9bY+I+cB23r7y1xXA/HRbB9zTxthmZtamdq4JfCgink3T/w/YC8wGVgD3pW73AVen6RXA/VF4kuLi8bNaTm5mZm2ZlHMAknqADwA7gO6IOARFkQAuSN1mA6/WPWwotZmZWQkUEe2tQOoC/gG4PSIelvRGREyvW340ImZI2gJ8PiKeSO3bgc9GxM5R61tHcYiI7u7uRf39/Q3HHR4epqur66TZdh041sYzm7jus+G1H3R0yHF1KtOC2dMm1L+Z96/TnKk5VcwE1cxVVqa+vr6dEdE7Xr+2Lgov6V3A14CvRMTDqfk1SbMi4lA6xHM4tQ8Bc+sePgc4OHqdEbEB2ADQ29sbtVqt4dgDAwOMtWzEmvVbmn4uk+HmBce5c1dbL+mk61Sm/atqE+rfzPvXac7UnCpmgmrmqmKmeu18CkjAvcDeiPjDukWbgNVpejXwSF37denTQEuAYyOHiszMrPPa+dPwQ8AngF2Snktt/w24A3hQ0lrgFeCatGwrcCUwCLwFfLKNsc3MrE0tF4B0LF9jLF7aoH8A17c6npmZTS5/E9jMLFPVOmNpp62eCZ5wv3nB8Uk5Sb//jo+2vQ6zXHkPwMwsUy4AZmaZcgEwM8uUC4CZWaZcAMzMMuUCYGaWKRcAM7NMuQCYmWXKBcDMLFMuAGZmmXIBMDPLlAuAmVmm/GNwdlqb6I/QncxEf6DOP0RnpzvvAZiZZarjBUDSckkvShqUtL7T45uZWaGjBUDSFOBLwBXAxcC1ki7uZAYzMyt0eg9gMTAYES9FxI+BfmBFhzOYmRmdPwk8G3i1bn4IuLTDGcwmxWSegB5LoxPTPvlsk0XFtdo7NJh0DbAsIj6V5j8BLI6IG+r6rAPWpdmLgBfHWN1M4PVTGLcVztS8KuZypuZUMRNUM1dZmX4uIt47XqdO7wEMAXPr5ucAB+s7RMQGYMN4K5L0TET0Tm689jhT86qYy5maU8VMUM1cVcxUr9PnAJ4G5kuaJ+kMYCWwqcMZzMyMDu8BRMRxSZ8GHgWmABsjYk8nM5iZWaHj3wSOiK3A1klY1biHiUrgTM2rYi5nak4VM0E1c1Ux00919CSwmZlVh38KwswsU6ddAej0T0lI2ijpsKTddW3nSdomaV+6n5HaJenulO15SQvrHrM69d8naXWbmeZKelzSXkl7JN1Ydi5JZ0l6StK3UqbfSe3zJO1I6/9qOvmPpDPT/GBa3lO3rltT+4uSlrWaqW59UyR9U9LmKmSStF/SLknPSXomtZW6TaX1TZf0kKRvp23rgyVvUxel12jk9s+SfrPs10rSb6VtfLekB9K2X/p23pKIOG1uFCeOvwtcCJwBfAu4+BSP+WFgIbC7ru0PgPVpej3whTR9JfB1QMASYEdqPw94Kd3PSNMz2sg0C1iYpt8NfIfipzVKy5XW3ZWm3wXsSGM9CKxM7X8K/Hqa/g3gT9P0SuCrafri9L6eCcxL7/eUNt/Dm4C/Ajan+VIzAfuBmaPaSt2m0jrvAz6Vps8AplchV1rvFOB7wM+VvJ3PBl4Gzq7bltaUvU21/Lp2esA2N4IPAo/Wzd8K3NqBcXs4sQC8CMxK07OAF9P0nwHXju4HXAv8WV37Cf0mId8jwEeqkgs4B3iW4lverwNTR79/FJ8E+2Canpr6afR7Wt+vxSxzgO3AZcDmNEbZmfbzzgJQ6nsHvIfiPzZVKVfdei4H/m/ZmXj71wzOS9vIZmBZ2dtUq7fT7RBQo5+SmF1Cju6IOASQ7i9I7WPlO2W50y7lByj+4i41VzrU8hxwGNhG8VfNGxFxvMH6fzp2Wn4MOH+yMwF/BHwW+Nc0f34FMgXwd5J2qvjmO5S/TV0IfB/43+lw2Z9LOrcCuUasBB5I06VliogDwBeBV4BDFNvITsrfplpyuhUANWir0seYxsp3SnJL6gK+BvxmRPxz2bki4icR8X6Kv7oXA794kvWf8kySPgYcjoid9c1lZko+FBELKX4V93pJHz5J305lmkpxqPOeiPgA8CbF4ZWyc5GOp18F/PV4XU91pnS+YQXFYZufBc6leB/HWn9H/0+YqNOtAIz7UxId8pqkWQDp/nBqHyvfpOeW9C6K//y/EhEPVyUXQES8AQxQHIedLmnk+yb16//p2Gn5NODIJGf6EHCVpP0Uvzx7GcUeQZmZiIiD6f4w8DcUxbLs924IGIqIHWn+IYqCUHYuKP6DfTYiXkvzZWb6FeDliPh+RPwL8DDwy5S8TbXqdCsAVfkpiU3AyCcJVlMcgx9pvy59GmEJcCztoj4KXC5pRvoL4vLU1hJJAu4F9kbEH1Yhl6T3Spqeps+m+IeyF3gc+PgYmUayfhx4LIqDoZuAlenTE/OA+cBTrWSKiFsjYk5E9FBsK49FxKoyM0k6V9K7R6YpXvPdlLxNRcT3gFclXZSalgIvlJ0ruZa3D/+MjF1WpleAJZLOSf8OR16n0raptnT6pEO7N4oz/d+hOL78uQ6M9wDFsb5/oajaaymO4W0H9qX781JfUVzw5rvALqC3bj2/Bgym2yfbzPQfKXYXnweeS7cry8wFvA/4Zsq0G/jvqf1Cig17kGIX/szUflaaH0zLL6xb1+dS1heBKybpfazx9qeASsuUxv5Wuu0Z2YbL3qbS+t4PPJPew7+l+MRM2dv6OcA/AdPq2srO9DvAt9N2/hcUn+SpxHY+0Zu/CWxmlqnT7RCQmZlNEhcAM7NMuQCYmWXKBcDMLFMuAGZmmXIBMDPLlAuAmVmmXADMzDL1/wEiw7CiFB57iQAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import numpy as np \n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "%matplotlib inline\n",
    "t1 = pd.read_table(\"temp/asyn_hemo.txt\", sep=\" \", \n",
    "                   names=(\"ID\", \"EVENT_ID\", \"ASYN\",\"HEMO\", \"DATE\"))\n",
    "t1.ASYN.hist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 180,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ID            int64\n",
       "EVENT_ID     object\n",
       "ASYN        float64\n",
       "HEMO         object\n",
       "DATE         object\n",
       "dtype: object"
      ]
     },
     "execution_count": 180,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "t1.dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "OK\n"
     ]
    }
   ],
   "source": [
    "def cut_hemo200(x):\n",
    "    if x == \"above\":\n",
    "        return 1\n",
    "    elif x == \"below\":\n",
    "        return 0\n",
    "    elif x == \">20\":\n",
    "        return 99\n",
    "    else:\n",
    "        try:\n",
    "            float(x)\n",
    "            return 1\n",
    "        except ValueError:\n",
    "            return 999\n",
    "t1['HEMO200'] = t1['HEMO'].apply(cut_hemo200)\n",
    "if len(t1[t1['HEMO200']>998]) == 0:\n",
    "    print(\"OK\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>EVENT_ID</th>\n",
       "      <th>ASYN</th>\n",
       "      <th>HEMO</th>\n",
       "      <th>DATE</th>\n",
       "      <th>HEMO200</th>\n",
       "      <th>ASYN_num</th>\n",
       "      <th>HEMO_num</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10874</td>\n",
       "      <td>BL</td>\n",
       "      <td>1778.3</td>\n",
       "      <td>148.0</td>\n",
       "      <td>2014-09-01</td>\n",
       "      <td>1</td>\n",
       "      <td>1778.3</td>\n",
       "      <td>148.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10874</td>\n",
       "      <td>V02</td>\n",
       "      <td>2459.3</td>\n",
       "      <td>below</td>\n",
       "      <td>2015-03-01</td>\n",
       "      <td>0</td>\n",
       "      <td>2459.3</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>10874</td>\n",
       "      <td>V04</td>\n",
       "      <td>1729.8</td>\n",
       "      <td>below</td>\n",
       "      <td>2015-08-01</td>\n",
       "      <td>0</td>\n",
       "      <td>1729.8</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>12224</td>\n",
       "      <td>BL</td>\n",
       "      <td>1029.7</td>\n",
       "      <td>below</td>\n",
       "      <td>2013-10-01</td>\n",
       "      <td>0</td>\n",
       "      <td>1029.7</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>12224</td>\n",
       "      <td>V02</td>\n",
       "      <td>891.2</td>\n",
       "      <td>below</td>\n",
       "      <td>2014-05-01</td>\n",
       "      <td>0</td>\n",
       "      <td>891.2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      ID EVENT_ID    ASYN   HEMO        DATE  HEMO200  ASYN_num  HEMO_num\n",
       "0  10874       BL  1778.3  148.0  2014-09-01        1    1778.3     148.0\n",
       "1  10874      V02  2459.3  below  2015-03-01        0    2459.3       NaN\n",
       "2  10874      V04  1729.8  below  2015-08-01        0    1729.8       NaN\n",
       "3  12224       BL  1029.7  below  2013-10-01        0    1029.7       NaN\n",
       "4  12224      V02   891.2  below  2014-05-01        0     891.2       NaN"
      ]
     },
     "execution_count": 183,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "t1['ASYN_num'] = pd.to_numeric(t1['ASYN'], errors=\"coerce\")\n",
    "t1['HEMO_num'] = pd.to_numeric(t1['HEMO'], errors=\"coerce\")\n",
    "t1.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
